File: Algorithm/Sort/program6-10.c
// Merge sort with recursive
#include <stdio.h>
#include <conio.h>
#define MAX 20

int arr[MAX];

void merge(int low, int mid, int high) {
    int temp[MAX];
    int i = low;
    int j = mid + 1;
    int k = low;

    while ((i <= mid) && (j <= high)) {
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        }
        else {
            temp[k++] = arr[j++];
        }
    }
    while (i <= mid) {
        temp[k++] = arr[i++];
    }
    while (j <= high) {
        temp[k++] = arr[j++];
    }
    
    for (i = low; i <= high; i++) {
        arr[i] = temp[i];
    }
}

void merge_sort(int low, int high) {
    int mid;
    if (low != high) {
        mid = (low + high) / 2;
        merge_sort(low, mid);
        merge_sort(mid + 1, high);
        merge(low, mid, high);
    }
}

void main() {
    int i, n;
    system("cls");
    printf("\nEnter the number of elements: ");
    scanf("%d", &n);
    for (i = 0; i < n; i++) {
        printf("\nEnter element %d: ", i + 1);
        scanf("%d", &arr[i]);
    }

    printf("\nUnsorted list is: \n");
    for (i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }

    merge_sort(0, n - 1);
    printf("\nSorted list is: \n");
    for (i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    getch();
}

File: Algorithm/Sort/program6-10.py
"""
Merge sort with recursive
Original: https://ithelp.ithome.com.tw/articles/10278179
"""

def merge(left, right):
    result = []

    while len(left) and len(right):
        if (left[0] < right[0]):
            result.append(left.pop(0))
        else:
            result.append(right.pop(0))
    result = result + left if len(left) else result + right
    return result

def merge_sort(arr):
    # Array只有1個元素 直接return
    if (len(arr) < 2):
        return arr
    
    mid = len(arr) // 2
    left_arr = arr[:mid]
    right_arr = arr[mid:]

    return merge(merge_sort(left_arr), merge_sort(right_arr))

if __name__ == "__main__":
    data = [54, 3, 12, 11, 7, 61]
    print(merge_sort(data))
